File: apps/api/src/schemas/domain.py
"""
Sentinel NER — Domain Schemas & Lifecycle Models (Stage 3)
Defines authoritative schemas for District, SlopeUnit, Road, RoadChainage,
Village, Asset, and LandslideEvent entities.
"""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, Generic, List, Optional, TypeVar, Union

from pydantic import BaseModel, ConfigDict, Field

from src.schemas.geojson import (
    GeoJSONLineString,
    GeoJSONMultiLineString,
    GeoJSONMultiPolygon,
    GeoJSONPoint,
    GeoJSONPolygon,
)

T = TypeVar("T")


# ==============================================================================
# ENUMS
# ==============================================================================

class EntityStatus(str, Enum):
    ACTIVE = "ACTIVE"
    INACTIVE = "INACTIVE"
    ARCHIVED = "ARCHIVED"


class SlopeUnitStatus(str, Enum):
    ACTIVE = "ACTIVE"
    MONITORED = "MONITORED"
    INACTIVE = "INACTIVE"


class RoadType(str, Enum):
    NATIONAL_HIGHWAY = "NATIONAL_HIGHWAY"
    STATE_HIGHWAY = "STATE_HIGHWAY"
    MAJOR_DISTRICT_ROAD = "MAJOR_DISTRICT_ROAD"
    RURAL_ROAD = "RURAL_ROAD"
    STRATEGIC_BORDER_ROAD = "STRATEGIC_BORDER_ROAD"


class RoadOperationalStatus(str, Enum):
    OPERATIONAL = "OPERATIONAL"
    RESTRICTED = "RESTRICTED"
    CLOSED = "CLOSED"
    UNDER_MAINTENANCE = "UNDER_MAINTENANCE"


class VillageStatus(str, Enum):
    ACTIVE = "ACTIVE"
    EVACUATED = "EVACUATED"
    INACCESSIBLE = "INACCESSIBLE"
    HISTORICAL = "HISTORICAL"


class AssetType(str, Enum):
    ROAD_INFRASTRUCTURE = "ROAD_INFRASTRUCTURE"
    BRIDGE = "BRIDGE"
    CULVERT = "CULVERT"
    RAILWAY = "RAILWAY"
    RAILWAY_STATION = "RAILWAY_STATION"
    POWER = "POWER"
    WATER = "WATER"
    TELECOM = "TELECOM"
    HEALTH = "HEALTH"
    SCHOOL = "SCHOOL"
    GOVERNMENT = "GOVERNMENT"
    OTHER = "OTHER"


class AssetOperationalStatus(str, Enum):
    OPERATIONAL = "OPERATIONAL"
    DAMAGED = "DAMAGED"
    DESTROYED = "DESTROYED"
    UNDER_REPAIR = "UNDER_REPAIR"
    OFFLINE = "OFFLINE"


class LandslideSourceType(str, Enum):
    FIELD_OBSERVATION = "FIELD_OBSERVATION"
    OFFICIAL_RECORD = "OFFICIAL_RECORD"
    IMPORTED_DATA = "IMPORTED_DATA"
    REMOTE_SENSING = "REMOTE_SENSING"
    OTHER = "OTHER"


class LandslideEventStatus(str, Enum):
    REPORTED = "REPORTED"
    VERIFIED = "VERIFIED"
    HISTORICAL = "HISTORICAL"
    REJECTED = "REJECTED"
    ARCHIVED = "ARCHIVED"


# ==============================================================================
# PAGINATION & QUERY PRIMITIVES
# ==============================================================================

class PaginatedResult(BaseModel, Generic[T]):
    items: List[T]
    total: int
    page: int
    limit: int
    pages: int
    has_next: bool
    has_prev: bool


# ==============================================================================
# 1. DISTRICT
# ==============================================================================

DistrictGeometry = Union[GeoJSONPolygon, GeoJSONMultiPolygon]


class DistrictCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    code: str = Field(..., min_length=2, max_length=20, description="Unique code e.g. 'MZ-AIZ'")
    state_code: str = Field(..., min_length=2, max_length=10, description="State code e.g. 'MZ'")
    state_name: str = Field(..., min_length=2, max_length=100)
    geometry: DistrictGeometry
    status: EntityStatus = EntityStatus.ACTIVE
    metadata: Optional[Dict[str, Any]] = None


class DistrictUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    state_name: Optional[str] = Field(None, min_length=2, max_length=100)
    geometry: Optional[DistrictGeometry] = None
    status: Optional[EntityStatus] = None
    metadata: Optional[Dict[str, Any]] = None


class DistrictResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    code: str
    state_code: str
    state_name: str
    geometry: Dict[str, Any]
    status: EntityStatus
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None


# ==============================================================================
# 2. SLOPE UNIT
# ==============================================================================

SlopeUnitGeometry = Union[GeoJSONPolygon, GeoJSONMultiPolygon]


class SlopeUnitCreate(BaseModel):
    code: str = Field(..., min_length=2, max_length=50, description="Unique terrain unit identifier")
    name: Optional[str] = Field(None, max_length=150)
    district_id: str = Field(..., description="ID of referencing district")
    state_code: str = Field(..., min_length=2, max_length=10)
    geometry: SlopeUnitGeometry
    area_sqkm: Optional[float] = Field(None, ge=0.0)
    status: SlopeUnitStatus = SlopeUnitStatus.ACTIVE
    metadata: Optional[Dict[str, Any]] = None


class SlopeUnitUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=150)
    geometry: Optional[SlopeUnitGeometry] = None
    area_sqkm: Optional[float] = Field(None, ge=0.0)
    status: Optional[SlopeUnitStatus] = None
    metadata: Optional[Dict[str, Any]] = None


class SlopeUnitResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    code: str
    name: Optional[str] = None
    district_id: str
    state_code: str
    geometry: Dict[str, Any]
    area_sqkm: Optional[float] = None
    status: SlopeUnitStatus
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None


# ==============================================================================
# 3. ROAD
# ==============================================================================

RoadGeometry = Union[GeoJSONLineString, GeoJSONMultiLineString]


class RoadCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=150)
    road_code: str = Field(..., min_length=2, max_length=50, description="Authoritative road code e.g. 'NH-54'")
    road_type: RoadType
    authority_organization_id: str = Field(..., description="Agency owning or managing road e.g. BRO or PWD")
    district_id: str = Field(...)
    state_code: str = Field(..., min_length=2, max_length=10)
    geometry: RoadGeometry
    operational_status: RoadOperationalStatus = RoadOperationalStatus.OPERATIONAL
    metadata: Optional[Dict[str, Any]] = None


class RoadUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=150)
    road_type: Optional[RoadType] = None
    operational_status: Optional[RoadOperationalStatus] = None
    geometry: Optional[RoadGeometry] = None
    metadata: Optional[Dict[str, Any]] = None


class RoadResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    road_code: str
    road_type: RoadType
    authority_organization_id: str
    district_id: str
    state_code: str
    geometry: Dict[str, Any]
    operational_status: RoadOperationalStatus
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None


# ==============================================================================
# 4. ROAD CHAINAGE
# ==============================================================================

class RoadChainageCreate(BaseModel):
    road_id: str = Field(..., description="ID of referencing road")
    chainage_km: float = Field(..., ge=0.0, description="Kilometer mark along road")
    geometry: GeoJSONPoint = Field(..., description="Location of chainage marker")
    district_id: str = Field(...)
    state_code: str = Field(..., min_length=2, max_length=10)
    metadata: Optional[Dict[str, Any]] = None


class RoadChainageUpdate(BaseModel):
    chainage_km: Optional[float] = Field(None, ge=0.0)
    geometry: Optional[GeoJSONPoint] = None
    metadata: Optional[Dict[str, Any]] = None


class RoadChainageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    road_id: str
    chainage_km: float
    geometry: Dict[str, Any]
    district_id: str
    state_code: str
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None


# ==============================================================================
# 5. VILLAGE
# ==============================================================================

VillageGeometry = Union[GeoJSONPoint, GeoJSONPolygon]


class VillageCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=150)
    village_code: Optional[str] = Field(None, max_length=50)
    district_id: str = Field(...)
    state_code: str = Field(..., min_length=2, max_length=10)
    geometry: VillageGeometry
    population: Optional[int] = Field(None, ge=0, description="Official census/record population if known, else null")
    status: VillageStatus = VillageStatus.ACTIVE
    metadata: Optional[Dict[str, Any]] = None


class VillageUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=150)
    village_code: Optional[str] = Field(None, max_length=50)
    geometry: Optional[VillageGeometry] = None
    population: Optional[int] = Field(None, ge=0)
    status: Optional[VillageStatus] = None
    metadata: Optional[Dict[str, Any]] = None


class VillageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    village_code: Optional[str] = None
    district_id: str
    state_code: str
    geometry: Dict[str, Any]
    population: Optional[int] = None
    status: VillageStatus
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None


# ==============================================================================
# 6. ASSET
# ==============================================================================

AssetGeometry = Union[GeoJSONPoint, GeoJSONLineString, GeoJSONPolygon]


class AssetCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=150)
    asset_type: AssetType
    organization_id: str = Field(..., description="ID of managing agency/department")
    district_id: str = Field(...)
    state_code: str = Field(..., min_length=2, max_length=10)
    geometry: AssetGeometry
    operational_status: AssetOperationalStatus = AssetOperationalStatus.OPERATIONAL
    metadata: Optional[Dict[str, Any]] = None


class AssetUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=150)
    asset_type: Optional[AssetType] = None
    operational_status: Optional[AssetOperationalStatus] = None
    geometry: Optional[AssetGeometry] = None
    metadata: Optional[Dict[str, Any]] = None


class AssetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    asset_type: AssetType
    organization_id: str
    district_id: str
    state_code: str
    geometry: Dict[str, Any]
    operational_status: AssetOperationalStatus
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None


# ==============================================================================
# 7. LANDSLIDE EVENT (Historical & Observational)
# ==============================================================================

LandslideGeometry = Union[GeoJSONPoint, GeoJSONPolygon]


class LandslideEventCreate(BaseModel):
    event_reference: str = Field(..., min_length=3, max_length=60, description="Unique event reference e.g. 'LS-2026-MZ-001'")
    event_time: datetime = Field(..., description="Time event occurred or was initiated")
    detected_time: Optional[datetime] = None
    reported_time: Optional[datetime] = None
    geometry: LandslideGeometry
    district_id: str = Field(...)
    state_code: str = Field(..., min_length=2, max_length=10)
    source: LandslideSourceType = LandslideSourceType.FIELD_OBSERVATION
    source_reference: Optional[str] = Field(None, max_length=150)
    status: LandslideEventStatus = LandslideEventStatus.REPORTED
    description: Optional[str] = Field(None, max_length=1000)
    evidence_references: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


class LandslideEventUpdate(BaseModel):
    status: Optional[LandslideEventStatus] = None
    description: Optional[str] = Field(None, max_length=1000)
    detected_time: Optional[datetime] = None
    reported_time: Optional[datetime] = None
    source_reference: Optional[str] = Field(None, max_length=150)
    evidence_references: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


class LandslideEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    event_reference: str
    event_time: datetime
    detected_time: Optional[datetime] = None
    reported_time: Optional[datetime] = None
    geometry: Dict[str, Any]
    district_id: str
    state_code: str
    source: LandslideSourceType
    source_reference: Optional[str] = None
    status: LandslideEventStatus
    description: Optional[str] = None
    evidence_references: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None

File: apps/web/src/lib/domain.ts
/**
 * Sentinel NER — Domain Data Typed Contracts (Stage 3)
 * Provides frontend TypeScript interfaces for authoritative operational entities
 * and typed API access functions.
 */

import { fetchFromAPI } from "./api";

// ==============================================================================
// GEOJSON TYPES
// ==============================================================================

export type GeoJSONGeometryType = "Point" | "LineString" | "Polygon" | "MultiPolygon" | "MultiLineString";

export interface GeoJSONPoint {
  type: "Point";
  coordinates: [number, number]; // [lng, lat]
}

export interface GeoJSONLineString {
  type: "LineString";
  coordinates: [number, number][];
}

export interface GeoJSONPolygon {
  type: "Polygon";
  coordinates: [number, number][][];
}

export interface GeoJSONMultiPolygon {
  type: "MultiPolygon";
  coordinates: [number, number][][][];
}

export type GeoJSONGeometry = GeoJSONPoint | GeoJSONLineString | GeoJSONPolygon | GeoJSONMultiPolygon;

// ==============================================================================
// DOMAIN ENTITY INTERFACES
// ==============================================================================

export interface District {
  id: string;
  name: string;
  code: string;
  state_code: string;
  state_name: string;
  geometry: GeoJSONPolygon | GeoJSONMultiPolygon;
  status: "ACTIVE" | "INACTIVE" | "ARCHIVED";
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface SlopeUnit {
  id: string;
  code: string;
  name?: string;
  district_id: string;
  state_code: string;
  geometry: GeoJSONPolygon | GeoJSONMultiPolygon;
  area_sqkm?: number;
  status: "ACTIVE" | "MONITORED" | "INACTIVE";
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface Road {
  id: string;
  name: string;
  road_code: string;
  road_type: "NATIONAL_HIGHWAY" | "STATE_HIGHWAY" | "MAJOR_DISTRICT_ROAD" | "RURAL_ROAD" | "STRATEGIC_BORDER_ROAD";
  authority_organization_id: string;
  district_id: string;
  state_code: string;
  geometry: GeoJSONLineString;
  operational_status: "OPERATIONAL" | "RESTRICTED" | "CLOSED" | "UNDER_MAINTENANCE";
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface RoadChainage {
  id: string;
  road_id: string;
  chainage_km: number;
  geometry: GeoJSONPoint;
  district_id: string;
  state_code: string;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface Village {
  id: string;
  name: string;
  village_code?: string;
  district_id: string;
  state_code: string;
  geometry: GeoJSONPoint | GeoJSONPolygon;
  population?: number;
  status: "ACTIVE" | "EVACUATED" | "INACCESSIBLE" | "HISTORICAL";
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface Asset {
  id: string;
  name: string;
  asset_type: "ROAD_INFRASTRUCTURE" | "BRIDGE" | "CULVERT" | "RAILWAY" | "RAILWAY_STATION" | "POWER" | "WATER" | "TELECOM" | "HEALTH" | "SCHOOL" | "GOVERNMENT" | "OTHER";
  organization_id: string;
  district_id: string;
  state_code: string;
  geometry: GeoJSONPoint | GeoJSONLineString | GeoJSONPolygon;
  operational_status: "OPERATIONAL" | "DAMAGED" | "DESTROYED" | "UNDER_REPAIR" | "OFFLINE";
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface LandslideEvent {
  id: string;
  event_reference: string;
  event_time: string;
  detected_time?: string;
  reported_time?: string;
  geometry: GeoJSONPoint | GeoJSONPolygon;
  district_id: string;
  state_code: string;
  source: "FIELD_OBSERVATION" | "OFFICIAL_RECORD" | "IMPORTED_DATA" | "REMOTE_SENSING" | "OTHER";
  source_reference?: string;
  status: "REPORTED" | "VERIFIED" | "HISTORICAL" | "REJECTED" | "ARCHIVED";
  description?: string;
  evidence_references?: string[];
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface PaginatedResult<T> {
  items: T[];
  total: number;
  page: number;
  limit: number;
  pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface APIEnvelope<T> {
  success: boolean;
  data: T;
  correlation_id: string;
  timestamp: string;
}

// ==============================================================================
// API HELPER ACCESSORS
// ==============================================================================

export async function fetchDistricts(params: { state_code?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<District>> {
  const query = new URLSearchParams();
  if (params.state_code) query.set("state_code", params.state_code);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<District>>>(`api/v1/districts?${query.toString()}`);
  return res.data;
}

export async function fetchSlopeUnits(params: { district_id?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<SlopeUnit>> {
  const query = new URLSearchParams();
  if (params.district_id) query.set("district_id", params.district_id);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<SlopeUnit>>>(`api/v1/slope-units?${query.toString()}`);
  return res.data;
}

export async function fetchRoads(params: { district_id?: string; authority_organization_id?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<Road>> {
  const query = new URLSearchParams();
  if (params.district_id) query.set("district_id", params.district_id);
  if (params.authority_organization_id) query.set("authority_organization_id", params.authority_organization_id);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<Road>>>(`api/v1/roads?${query.toString()}`);
  return res.data;
}

export async function fetchLandslideEvents(params: { district_id?: string; status?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<LandslideEvent>> {
  const query = new URLSearchParams();
  if (params.district_id) query.set("district_id", params.district_id);
  if (params.status) query.set("status", params.status);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<LandslideEvent>>>(`api/v1/landslide-events?${query.toString()}`);
  return res.data;
}

export async function fetchRoadChainages(params: { road_id?: string; district_id?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<RoadChainage>> {
  const query = new URLSearchParams();
  if (params.road_id) query.set("road_id", params.road_id);
  if (params.district_id) query.set("district_id", params.district_id);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<RoadChainage>>>(`api/v1/road-chainages?${query.toString()}`);
  return res.data;
}

export async function fetchVillages(params: { district_id?: string; status?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<Village>> {
  const query = new URLSearchParams();
  if (params.district_id) query.set("district_id", params.district_id);
  if (params.status) query.set("status", params.status);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<Village>>>(`api/v1/villages?${query.toString()}`);
  return res.data;
}

export async function fetchAssets(params: { district_id?: string; organization_id?: string; asset_type?: string; page?: number; limit?: number } = {}): Promise<PaginatedResult<Asset>> {
  const query = new URLSearchParams();
  if (params.district_id) query.set("district_id", params.district_id);
  if (params.organization_id) query.set("organization_id", params.organization_id);
  if (params.asset_type) query.set("asset_type", params.asset_type);
  if (params.page) query.set("page", params.page.toString());
  if (params.limit) query.set("limit", params.limit.toString());

  const res = await fetchFromAPI<APIEnvelope<PaginatedResult<Asset>>>(`api/v1/assets?${query.toString()}`);
  return res.data;
}

export interface SpatialQueryResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  districts?: District[];
  slope_units?: SlopeUnit[];
  roads?: Road[];
  road_chainages?: RoadChainage[];
  villages?: Village[];
  assets?: Asset[];
  landslide_events?: LandslideEvent[];
}

export async function fetchSpatialNearby(params: {
  lon: number;
  lat: number;
  radius_meters?: number;
  layers?: string;
  entity_type?: string;
}): Promise<SpatialQueryResponse> {
  const query = new URLSearchParams();
  query.set("longitude", params.lon.toString());
  query.set("latitude", params.lat.toString());
  if (params.radius_meters) query.set("radius_meters", params.radius_meters.toString());
  if (params.layers) query.set("layers", params.layers);
  if (params.entity_type) query.set("entity_type", params.entity_type);

  const res = await fetchFromAPI<APIEnvelope<SpatialQueryResponse>>(`api/v1/spatial/nearby?${query.toString()}`);
  return res.data;
}

export async function fetchSpatialPointInGeometry(params: {
  lon: number;
  lat: number;
  layers?: string;
}): Promise<SpatialQueryResponse> {
  const query = new URLSearchParams();
  query.set("longitude", params.lon.toString());
  query.set("latitude", params.lat.toString());
  if (params.layers) query.set("layers", params.layers);

  const res = await fetchFromAPI<APIEnvelope<SpatialQueryResponse>>(`api/v1/spatial/point-in-geometry?${query.toString()}`);
  return res.data;
}

export async function fetchSpatialBBox(params: {
  min_lng: number;
  min_lat: number;
  max_lng: number;
  max_lat: number;
  layers?: string;
}): Promise<SpatialQueryResponse> {
  const query = new URLSearchParams();
  query.set("min_lng", params.min_lng.toString());
  query.set("min_lat", params.min_lat.toString());
  query.set("max_lng", params.max_lng.toString());
  query.set("max_lat", params.max_lat.toString());
  if (params.layers) query.set("layers", params.layers);

  const res = await fetchFromAPI<APIEnvelope<SpatialQueryResponse>>(`api/v1/spatial/bbox?${query.toString()}`);
  return res.data;
}

// ==============================================================================
// NORTH EAST REGION (NER) STATES & DISTRICTS
// ==============================================================================

export interface NERStateDistrictGroup {
  stateCode: string;
  stateName: string;
  districts: {
    id: string;
    name: string;
    code: string;
  }[];
}

export const NER_STATE_GROUPS: NERStateDistrictGroup[] = [
  {
    stateCode: "MZ",
    stateName: "Mizoram",
    districts: [
      { id: "dst-aizawl", name: "Aizawl District", code: "MZ-AIZ" },
      { id: "dst-kolasib", name: "Kolasib District", code: "MZ-KOL" },
      { id: "dst-champhai", name: "Champhai District", code: "MZ-CHA" },
      { id: "dst-lunglei", name: "Lunglei District", code: "MZ-LUN" },
      { id: "dst-serchhip", name: "Serchhip District", code: "MZ-SER" },
    ],
  },
  {
    stateCode: "AS",
    stateName: "Assam",
    districts: [
      { id: "dst-dima-hasao", name: "Dima Hasao District", code: "AS-DH" },
      { id: "dst-kamrup-metro", name: "Kamrup Metropolitan", code: "AS-KM" },
      { id: "dst-cachar", name: "Cachar District", code: "AS-CA" },
    ],
  },
  {
    stateCode: "ML",
    stateName: "Meghalaya",
    districts: [
      { id: "dst-east-khasi-hills", name: "East Khasi Hills (Shillong)", code: "ML-EKH" },
      { id: "dst-west-jaintia-hills", name: "West Jaintia Hills", code: "ML-WJH" },
    ],
  },
  {
    stateCode: "AR",
    stateName: "Arunachal Pradesh",
    districts: [
      { id: "dst-papum-pare", name: "Papum Pare (Itanagar)", code: "AR-PP" },
      { id: "dst-west-kameng", name: "West Kameng District", code: "AR-WK" },
      { id: "dst-tawang", name: "Tawang District", code: "AR-TW" },
    ],
  },
  {
    stateCode: "MN",
    stateName: "Manipur",
    districts: [
      { id: "dst-imphal-west", name: "Imphal West District", code: "MN-IW" },
      { id: "dst-noney", name: "Noney (Tupul Corridor)", code: "MN-NN" },
      { id: "dst-churachandpur", name: "Churachandpur District", code: "MN-CC" },
    ],
  },
  {
    stateCode: "NL",
    stateName: "Nagaland",
    districts: [
      { id: "dst-kohima", name: "Kohima District", code: "NL-KO" },
      { id: "dst-dimapur", name: "Dimapur District", code: "NL-DI" },
      { id: "dst-mokokchung", name: "Mokokchung District", code: "NL-MK" },
    ],
  },
  {
    stateCode: "SK",
    stateName: "Sikkim",
    districts: [
      { id: "dst-gangtok", name: "Gangtok District", code: "SK-GT" },
      { id: "dst-namchi", name: "Namchi District", code: "SK-NM" },
      { id: "dst-mangan", name: "Mangan District", code: "SK-MG" },
    ],
  },
  {
    stateCode: "TR",
    stateName: "Tripura",
    districts: [
      { id: "dst-west-tripura", name: "West Tripura (Agartala)", code: "TR-WT" },
      { id: "dst-dhalai", name: "Dhalai District", code: "TR-DH" },
    ],
  },
];

